### DISCUSSION TITLE
Heat boosting with a valve - followup

### DISCUSSION THREAD

**Discussion by antamy** [tagyoureit/nodejs-poolController#741] [App and O/S Config Questions]
This is in followup to my question from February (https://github.com/tagyoureit/nodejs-poolController/discussions/687#discussioncomment-4950270), where I want to create a "Heat Boost" mode on my pool such that when the solar heating is on the heater is also turned on.  We have determined that by hooking into Valve B on my intelliTouch i5+3, a 24V relay can be activated which operates the fireman's switch for the heater.

I only want this to occur when (a) the boost function is activated via the panel or the dashPanel UI and (b) the solar heating is operational.   I have created two features - HEATBOOST and BOOSTRELAY.  BOOSTRELAY is assigned to ValveB.  HEATBOOST is currently just a custom name, but it appears on the panel and the UI.

Using server:4200/state/all, I see the ids for the features, but I am not sure where to look to find the API that I use to determine the true / false value of HEATBOOST and to insert it into my modified version of the automation code provided by @rstrouse in the original thread.  What do I replace <<heatBoostFeatureId == 'true">> with in the following:

```
{
  {
  "context": {
    "name": "poolRules",
    "options": {},
    "vars": {
      "boostRelayFeatureId": 13,
      "heatBoostFeatureId": 41
    }
  },
  "events": [
    {
      "name": "Heat Boost",
      "description": "Use heater when solar is active and HeatBoost is ON",
      "processor": [
        "try {",
        "let heaterOn = false;",
        "let body = data.bodies[0];",
        "if(typeof body.heatStatus !== 'undefined' && body.heatStatus.name === 'solar' && <<heatBoostFeatureId == 'true">> ) heaterOn = true;",
        "await sys.board.circuits.setFeatureStateAsync(vars.boostRelayFeatureId, heaterOn);",
        "} catch(err) { logger.error(`Error triggering heat boost feature ${err.message}`); }"
      ]
    }
  ]
}


```

**Comment by rstrouse** [comment]
So in this instance the gas heater is not connected to the gas heater relay on the IntelliTouch at all?

You should know that the there is no such event named `Heat Boost` so your code will never be called with that.  The event name you are wanting to hook is `temps`

**Reply by antamy** [reply]
The heater is connected and I have to play around with the board to connect my additional relay in.  

I'm still getting up to speed with the syntax and think that the following will work, just not sure if the setFeatureStateAsync is the correct call to make..  I did upload a bindings file, but now can't upload another because the file already exists.  But I can't find it - where are they stored?

```
{
  "context": {
    "name": "poolRules",
    "options": {},
    "vars": {
      "poolBodyId": 0,
      "heatBoostFeatureId": 41,
      "boostRelayFeatureId": 42
    }
  },
  "events": [
    {
      "name": "temps",
      "description": "Use heater when solar is active and HeatBoost is ON",
      "processor": [
        "try {",
        "let boostRelayOn = false;",
        "let body = data.bodies[poolBodyId];",
        "let feature = data.features[heatBoostFeatureId];",

        "if(typeof body.heatStatus !== 'undefined' && (body.heatStatus.name == 'solar' && feature.isOn == 'true')) boostRelayOn = true;",
        "}",
        "await sys.board.circuits.setFeatureStateAsync(vars.boostRelayFeatureId, boostRelayOn);",
        "} catch(err) { logger.error(`Error triggering heat boost feature ${err.message}`); }"
      ]
    }
  ]
}
```

**Reply by antamy** [reply]
I'm kind of tearing my hair out with this, battling against the syntax and data model that I don't seem to understand.  I have a features rule loaded that fires when I click HeatBoost on the UI:


```
{
  "context": {
    "name": "heatBoostRule",
    "options": {},
    "vars": {
      "heatBoostFeatureId": 41,
      "boostRelayFeatureId": 42
    }
  },
  "events": [
    {
      "name": "feature",
      "description": "Look for HeatBoost feature and act accordingly",
      "processor": [
        "try {",
          "logger.info('Processing HeatBoost');",

          "let boostRelayOn = false;",
          "let feature = data.features[heatBoostFeatureId];",

          "for (let i = 0; i < data.bodies.length; i++) {",

            "let body = data.bodies[i];",

            "if(typeof body.heatStatus !== 'undefined' && (body.heatStatus.name == 'solar' && feature.isOn == 'true')) {",
               "boostRelayOn = true;",
               "logger.info ('Solar is on, HeatBoost is on, activating ...');",
            "}",
          "}",

          "await sys.board.circuits.setFeatureStateAsync(vars.boostRelayFeatureId, boostRelayOn);",
        "} catch(err) { logger.error(`Error triggering heat boost feature ${err.message}`); }"
      ]
    }
  ]
}
```



but it fails with:

```
[26/04/2023, 15:24:23] info: Processing HeatBoost
[26/04/2023, 15:24:23] error: Error triggering heat boost feature heatBoostFeatureId is not defined
```

Is my syntax incorrect?  Am I referencing the feature incorrectly (I took the ids from /state/features where there are two features with ids 41 and 42 respectively.

Thanks

**Comment by tagyoureit** [comment]
Vars need to be accessed via a special context variable that will be loaded internally.  Use `ctx.vars.[varname]`.

`"let feature = data.features[ctx.vars.heatBoostFeatureId];",`

Same for boostRelay, of course.

**Reply by antamy** [reply]
I tried that and got 

```
[26/04/2023, 20:17:57] info: Processing HeatBoost
[26/04/2023, 20:17:57] error: Error triggering heat boost feature ctx is not defined
```

I also tried just replacing these variables with the actual numbers and get : 

```
[26/04/2023, 20:37:27] info: Processing HeatBoost
[26/04/2023, 20:37:27] error: Error triggering heat boost feature Cannot read properties of undefined (reading '41')
```
Exasperating!

**Reply by rstrouse** [reply]
Let's get you a little closer.

```
{
  "context": {
    "name": "heatBoostRule",
    "options": {},
    "vars": {
      "heatBoostFeatureId": 41,
      "boostRelayFeatureId": 42
    }
  },
  "events": [
    {
      "name": "feature",
      "description": "Look for HeatBoost feature and act accordingly",
      "processor": [
        "try {",
        "logger.info('Processing HeatBoost');",

        "let boostRelayOn = false;",
        "let feature = state.features.getItemById(vars.heatBoostFeatureId, false);",
        "for (let i = 0; i < data[0].bodies.length; i++) {",
        "let body = data[0].bodies[i];",
        "if(typeof body.heatStatus !== 'undefined' && (body.heatStatus.name === 'solar' && feature.isOn === 'true')) {",
        "boostRelayOn = true;",
        "logger.info ('Solar is on, HeatBoost is on, activating ...');",
        "}",
        "}",

        "await sys.board.circuits.setFeatureStateAsync(vars.boostRelayFeatureId, boostRelayOn);",
        "} catch(err) { logger.error(`Error triggering heat boost feature ${err.message}`); }"
      ]
    }
  ]
}
```

**Reply by rstrouse** [reply]
I just looked at this again and realized you are now hooking the feature event not the temps event.   This means that this event will be triggered each time the feature is turned on or off.  Also, the data sent with this event does not include bodies, it includes state information for the feature.

What is it that you are trying to do?  Let's work this out like this.

First what are our triggers.  What are the things that will cause us to perform an action.  Then what are our actions.

**Reply by antamy** [reply]
Thanks.  I tried it, however it doesn't appear to like data[0].bodies.length.  

`[26/04/2023, 21:01:21] error: Error triggering heat boost feature Cannot read properties of undefined (reading 'length')`

I am trying to implement an on/off button for boosting the heat.  When ON, I want it to activate Valve B, but only if solar is ON.  When OFF I want it to de-activate Valve B.   I realize using the feature event doesn't cater for when solar turns off, so I would probably need a temps event to handle that.

I tried this with the temps event, but it doesn't fire when I turn on the Heat Boost.  That's why I went down the feature path.  I just assumed that all the information in the state model is accessible.

I have two virtual circuits - HEATBOOST and BOOSTRELAY.  HEATBOOST isn't "connected" to anything, BOOSTRELAY is connected to Valve B.

**Reply by rstrouse** [reply]
It doesn't like bodies because that is only valid data on the temps event.  Not on the feature event.  You can see these events in the console in dashPanel.  It sounds like you want it to fire in both instances.

So let me see if I can get this straight.

Whenever...
1. There is a call for solar heat
2. AND the boost feature (boostFeatureId) is turned on

Then...
1. Divert the heat boost valve relay (boostRelayFeatureId) by turning it on.

Otherwise...
1. Turn the heat boost valve relay off

**Reply by antamy** [reply]
I think your summary is correct.  Not sure of the nuance of a "call for solar heat", but if it means if solar heating is active, then yes.  This is also why I created HEATBOOST and BOOSTRELAY, so that the relay would only turn on when HEATBOOST and solar are on.

So, is there a way to trigger a temps event from a feature event?  Then when HEATBOOST turns ON, I could fire a temps event.  Are features available in a temps event?

Where is the console on dashPanel?

**Reply by rstrouse** [reply]
Here you go.

```
{
  "context": {
    "name": "heatBoostRule",
    "options": {},
    "vars": {
      "heatBoostFeatureId": 41,
      "boostRelayFeatureId": 42
    }
  },
  "events": [
    {
      "name": "temps",
      "description": "Look for HeatBoost feature and act accordingly",
      "processor": [
        "try {",
        "logger.info('Processing HeatBoost');",
        "let boostRelayOn = false;",
        "let feature = state.features.getItemById(vars.heatBoostFeatureId, false);",
        "let relay = state.features.getItemById(vars.boostRelayFeatureId, false);",
        "for (let i = 0; i < data[0].bodies.length; i++) {",
        "let body = data[0].bodies[i];",
        "if(typeof body.heatStatus !== 'undefined' && (body.heatStatus.name === 'solar' && feature.isOn)) {",
        "boostRelayOn = true;",
        "if(!relay.isOn) logger.info ('Solar is on, HeatBoost is on, activating ...');",
        "}",
        "}",
        "await sys.board.features.setFeatureStateAsync(vars.boostRelayFeatureId, boostRelayOn);",
        "} catch(err) { logger.error(`Error triggering heat boost feature ${err.message}`); }"
      ]
    }
  ]
}

```

**Reply by antamy** [reply]
Thanks, I will try this tomorrow when the sun is out!

When does the temps event get fired?  If solar heating is active and then I turn HEATBOOST on, that wouldn't fire the temps event would it?  Would the relay only turn on when the temperature changes?

**Reply by rstrouse** [reply]
The temps event gets fired on all edge cases where there would be a change related to the heater to include setpoints, status, temp changes... etc.  This is the only event you need and hide the relay id.

**Reply by antamy** [reply]
This appears to be working, thank you.  There is a small delay between toggling HEATBOOST in the UI and BOOSTRELAY / Valve B being toggled, but no big issue there.   I didn't bring the correct connector with me to connect to the Valve B outlet on the board, so I have to track that down and install the hardware.

**Comment by rstrouse** [comment]
If you use the one below this will also trigger immediately when the feature is clicked on or off.  If the pool circuit is first engaged or disengaged there will be a delay as the heater functions are geared up and will not rotate until the solar valve is rotated.  This will help to stabilize flow as the pump ramps.

Essentially I added an event when you click the Heat Boost feature.

```
{
  "context": {
    "name": "heatBoostRule",
    "options": {},
    "vars": {
      "heatBoostFeatureId": 41,
      "boostRelayFeatureId": 42
    }
  },
  "events": [
    {
      "name": "temps",
      "description": "Look for HeatBoost feature and act accordingly",
      "processor": [
        "try {",
        "let boostRelayOn = false;",
        "let feature = state.features.getItemById(vars.heatBoostFeatureId, false);",
        "let relay = state.features.getItemById(vars.boostRelayFeatureId, false);",
        "for (let i = 0; i < data[0].bodies.length; i++) {",
        "let body = data[0].bodies[i];",
        "if(typeof body.heatStatus !== 'undefined' && (body.heatStatus.name === 'solar' && feature.isOn)) {",
        "boostRelayOn = true;",
        "if(!relay.isOn) logger.info ('Solar is on, HeatBoost is on, activating ...');",
        "}",
        "}",
        "await sys.board.features.setFeatureStateAsync(vars.boostRelayFeatureId, boostRelayOn);",
        "} catch(err) { logger.error(`Error triggering heat boost feature ${err.message}`); }"
      ]
    },
    {
      "name": "feature",
      "filter": "@bind=data.id; === @bind=vars.heatBoostFeatureId;",
      "description": "Trigger the HeatBoost feature immediately",
      "processor": [
        "try {",
        "let feature = data[0];",
        "let boostRelayOn = false;",
        "if(feature.isOn) {",
        "for(let i = 0; i < state.temps.bodies.length; i++) {",
        "let body = state.temps.bodies.getItemByIndex(i).getExtended();",
        "if(typeof body.heatStatus !== 'undefined' && body.heatStatus.name === 'solar') boostRelayOn = true;",
        "}",
        "}",
        "await sys.board.features.setFeatureStateAsync(vars.boostRelayFeatureId, boostRelayOn);",
        "} catch(err) { logger.error(`Error triggering heat boost feature ${err.message}`); }"
      ]
    }
  ]
}
```

**Reply by antamy** [reply]
I will try that too, tomorrow.  It's already too hot today!

I have to find somewhere to learn some of this syntax.   "filter": "@bind=data.id; === @bind=vars.heatBoostFeatureId;",   is a mystery to me!

**Reply by antamy** [reply]
ok, I finally got the hardware installed and there is still an issue.  I think the script overall is fine, however it toggles a feature (BOOSTRELAY) on or off.  That doesn't actually turn on Valve B.

I have Valve B associated with BOOSTRELAY on my controller.  In njspc settings > Circuits, BOOSTRELAY appears under both Aux-Circuits and Features.  If I put Features BOOSTRELAY on the dashPanel and toggle it on or off, it doesn't affect Valve B.  If I put Aux-Circuits BOOSTRELAY on dashPanel, that does control Valve B.

How would this line in the script need to be modified to control the circuit?

`  "await sys.board.features.setFeatureStateAsync(vars.boostRelayFeatureId, boostRelayOn);",`

What I see in the log when I toggle Aux-Circuits BOOSTRELAY is

```
[16/05/2023, 15:03:59] info: [15:03:59] 10.10.20.73 PUT /state/circuit/setState {"id":5,"state":true}
[16/05/2023, 15:04:02] info: [15:04:02] 10.10.20.73 PUT /state/circuit/setState {"id":5,"state":false}
```

Thanks!

**Reply by rstrouse** [reply]
Set the boostRelayFeatureId to 5 in the file and change the line to.

`await sys.board.circuits.setCircuitStateAsync(vars.boostRelayFeatureId, boostRelayOn);`

That being said I would guess that the feature ids at the top of the file are not correct.

**Reply by antamy** [reply]
I'll try it tomorrow.  I checked the feature ids, they are correct.

**Reply by rstrouse** [reply]
Oh and change the event to be `circuit` instead of `feature` for second event.
